src/body_parser.rs

```diff
@@ -0,0 +1,124 @@
+use hyper::header::ContentType;
+use hyper::mime::{Mime, SubLevel, TopLevel};
+use serialize::{Decodable, json};
+use request::Request;
+use plugin::{Plugin, Pluggable};
+use status::StatusCode;
+use std::error::Error as StdError;
+use std::fmt;
+use std::io::{self, ErrorKind, Read};
+use typemap::Key;
+use urlencoded::{self, Query};
+
+struct BodyReader;
+
+impl Key for BodyReader {
+    type Value = String;
```

Comment:
I think this might be more useful in future to be `Vec<u8>` and be public, but I'm happy to leave it like this as we should probably just clean up the `Request` api before making any more externalities based on it.